zipkin/src/main/java/zipkin/storage/InMemorySpanStore.java

```diff
@@ -167,31 +166,18 @@ synchronized int evictToRecoverSpans(int spansToRecover) {
   }
 
   /** Returns the count of spans evicted. */
-  private synchronized int deleteOldestTrace() {
+  private int deleteOldestTrace() {
     int spansEvicted = 0;
     long traceId = spansByTraceIdTimeStamp.delegate.lastKey()._1;
-    long latestTimeStamp = 0; // for ALL services names encountered
     Collection<Pair<Long>> traceIdTimeStamps = traceIdToTraceIdTimeStamps.remove(traceId);
-    LinkedHashSet<String> serviceNames = new LinkedHashSet<>();
     for (Iterator<Pair<Long>> traceIdTimeStampIter = traceIdTimeStamps.iterator();
       traceIdTimeStampIter.hasNext(); ) {
       Pair<Long> traceIdTimeStamp = traceIdTimeStampIter.next();
-      latestTimeStamp = Math.max(latestTimeStamp, traceIdTimeStamp._2);
       Collection<Span> spans = spansByTraceIdTimeStamp.remove(traceIdTimeStamp);
       spansEvicted += spans.size();
-      for (Iterator<Span> spansIter = spans.iterator(); spansIter.hasNext(); ) {
-        Span span = spansIter.next();
-        if (span.timestamp != null) {
-          latestTimeStamp = Math.max(latestTimeStamp, span.timestamp);
```

Comment:
note with this change, we no longer need to derive data from spans when cleaning up. Instead we use the values pinned at lookup time (because we are deleting by trace ID, this all works)